libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,557 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Stack layout generator for Yul to EVM code generation.
+ */
+
+#include <libyul/backends/evm/StackLayoutGenerator.h>
+
+#include <libyul/backends/evm/StackHelpers.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/cxx20.h>
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/concat.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace std;
+
+StackLayoutGenerator::StackLayoutGenerator(StackLayout& _layout, vector<VariableSlot> _currentFunctionReturnVariables):
+	m_layout(_layout),
+	m_currentFunctionReturnVariables(move(_currentFunctionReturnVariables))
+{
+}
+
+namespace
+{
+struct PreviousSlot { size_t slot; };
+
+template<typename Callable>
+Stack createIdealLayout(Stack const& _post, vector<variant<PreviousSlot, StackSlot>> _layout, Callable _generateSlotOnTheFly)
+{
+	if (_layout.empty())
+		return Stack{};
+
+	struct ShuffleOperations
+	{
+		vector<variant<PreviousSlot, StackSlot>>& layout;
+		Stack const& post;
+		std::set<StackSlot> outputs;
+		std::map<StackSlot, int> multiplicity;
+		Callable generateSlotOnTheFly;
+		ShuffleOperations(
+			vector<variant<PreviousSlot, StackSlot>>& _layout,
+			Stack const& _post,
+			Callable _generateSlotOnTheFly
+		): layout(_layout), post(_post), generateSlotOnTheFly(_generateSlotOnTheFly)
+		{
+			for (auto const& layoutSlot: layout)
+				if (StackSlot const* slot = get_if<StackSlot>(&layoutSlot))
+					outputs.insert(*slot);
+
+			for (auto const& layoutSlot: layout)
+				if (StackSlot const* slot = get_if<StackSlot>(&layoutSlot))
+					--multiplicity[*slot];
+			for (auto&& slot: post)
+				if (outputs.count(slot) || generateSlotOnTheFly(slot))
+					++multiplicity[slot];
+		}
+		bool isCompatible(size_t _source, size_t _target)
+		{
+			return
+				_source < layout.size() &&
+				_target < post.size() &&
+				(
+					std::holds_alternative<JunkSlot>(post.at(_target)) ||
+					std::visit(util::GenericVisitor{
+						[&](PreviousSlot const&) {
+							return !outputs.count(post.at(_target)) && !generateSlotOnTheFly(post.at(_target));
+						},
+						[&](StackSlot const& _s) { return _s == post.at(_target); }
+					}, layout.at(_source))
+				);
+		}
+		bool sourceIsSame(size_t _lhs, size_t _rhs)
+		{
+			return std::visit(util::GenericVisitor{
+				[&](PreviousSlot const&, PreviousSlot const&) { return true; },
+				[&](StackSlot const& _lhs, StackSlot const& _rhs) { return _lhs == _rhs; },
```

Comment:
Any two `FunctionReturnLabelSlot` would be the same. Why should they be the same here?

Reply:
There is no overlap between functions, i.e. there is no situation in which there can be two ``FunctionReturnLabelSlot``s referring to the return labels of different functions on the stack. (Maybe I should clarify that "complete stack" really means complete relative to the base of each function, as @chriseth already wanted clarified elsewhere.)
Or put differently: any ``FunctionReturnLabelSlot`` on stack has the same value, which is always the value the next ``FunctionReturn`` is supposed to jump to. In that sense, if it was to occur twice on the stack (it won't ever happen, but still), they'd be copies of each other.

Reply:
> There is no overlap between functions, i.e. there is no situation in which there can be two FunctionReturnLabelSlots referring to the return labels of different functions on the stack

I still think that we should distinguish between them. It's not obvious to me that there wouldn't be any overlap, especially since it can be freely generated at any point.

Reply:
Let me see if I can store a reference to the function in ``FunctionReturnLabelSlot`` and just assert in the comparison operators that the compared slots always refer to the same function, that should address the issue and is probably safer than really allowing distinguishing them.

Reply:
Done.